src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -612,8 +615,57 @@ public virtual ProtoCore.Type GetTypeHintForOutput(int index)
         {
              return ProtoCore.TypeSystem.BuildPrimitiveTypeObject(ProtoCore.PrimitiveType.kTypeVar);
         }
-        #endregion
 
+      
+        /// <summary>
+        /// Gets or sets a value indicating whether the node is frozen.
+        /// If the node is set to freeze by the user, then Freeze always true.
+        /// Otherwise, if any of the node has input node frozen, then Freeze is true.        
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if this node is frozen; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsFrozen
+        {
+            get
+            {
+                var gatheredInputFrozenNodes = new List<NodeModel>();
+                CheckIfAnyUpstreamNodeIsFrozen(this, gatheredInputFrozenNodes);
```

Comment:
Can we combine this method and the check below? You don't actually need the list of frozen nodes. How about one method called 
```c# 
bool IsAnyUpstreamFrozen()
```
Then you can just say:

```c#
return IsAnyUpstreamFrozen()? true : explicitFrozen
```